### PR TITLE
HeaderValueMatch support custom descriptor key

### PR DESCRIPTION
**Pull request by zirain**
Signed-off-by: hejianpeng <hejianpeng2@huawei.com>

before:
```yaml
- header_value_match:
      descriptor_value: "svc1"
      expect_match: true
      headers:
        - name: :path
          prefix_match: /svc1
```
```console
[2022-03-14 13:45:02.793][3414241][debug][filter] [external/envoy/source/extensions/filters/http/local_ratelimit/local_ratelimit.cc:120] populate descriptors: key=request_method value=GET
[2022-03-14 13:45:02.793][3414241][debug][filter] [external/envoy/source/extensions/filters/http/local_ratelimit/local_ratelimit.cc:120] populate descriptors: key=header_match value=svc1
```

after:
```
- header_value_match:
      descriptor_key: path_match
      descriptor_value: "svc1"
      expect_match: true
      headers:
        - name: :path
          prefix_match: /svc1
```

```console
[2022-03-14 13:45:35.923][3414669][debug][filter] [external/envoy/source/extensions/filters/http/local_ratelimit/local_ratelimit.cc:120] populate descriptors: key=request_method value=GET
[2022-03-14 13:45:35.923][3414669][debug][filter] [external/envoy/source/extensions/filters/http/local_ratelimit/local_ratelimit.cc:120] populate descriptors: key=path_match value=svc1
```

Commit Message: HeaderValueMatch support custom descriptor key
Additional Description:
Risk Level: low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/20249
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by adisuissa — CHANGES_REQUESTED** [review]
Overall LGTM, but needs tests, and updated release notes.
API LGTM.
/wait

**Comment by zirain** [comment]
@adisuissa can you rerun the pipeline? I think there's nothing special with tsan.

**Comment by adisuissa** [comment]
Per the contributions guidelines, please avoid force-pushing as it is hard to track changes.

**Comment by zirain** [comment]
> Per the contributions guidelines, please avoid force-pushing as it is hard to track changes.

ok, I will not do that next time.

**Comment by zirain** [comment]
update tests and release-notes, PTAL @adisuissa

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks!
/assign-from @envoyproxy/senior-maintainers

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -25,6 +25,7 @@ Minor Behavior Changes
 * http: when writing custom filters, `injectEncodedDataToFilterChain` and `injectDecodedDataToFilterChain` now trigger sending of headers if they were not yet sent due to `StopIteration`. Previously, calling one of the inject functions in that state would trigger an assertion. See issue #19891 for more details.
 * listener: the :ref:`ipv4_compat <envoy_api_field_core.SocketAddress.ipv4_compat>` flag can only be set on Ipv6 address and Ipv4-mapped Ipv6 address. A runtime guard is added ``envoy.reloadable_features.strict_check_on_ipv4_compat`` and the default is true.
 * perf: ssl contexts are now tracked without scan based garbage collection and greatly improved the performance on secret update.
+* ratelimit: header_value_match support custom descriptor key.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:28]: please ref link to the field

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small release note comment.

/wait

File: docs/root/version_history/current.rst (hunk 1 above)
Comment (zirain) [reply docs/root/version_history/current.rst]: updated

Comment (zirain) [reply docs/root/version_history/current.rst]: updated, sorry for the mistake about miss signoff, which cause me do another force-pushing

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**